packages/client/src/components/SafeContacts.js

```diff
@@ -1,43 +1,191 @@
+    </div>
+  );
+}
+
+function RemoveContactModal({ contact, onConfirm, confirmText, headerTitle }) {
+  const modalContext = useModalContext();
+  return (
+    <div className="py-5 has-text-black">
+      <ModalHeader className="px-5 pb-5 border-light-bottom" title={headerTitle} />
+      <div className="column is-flex is-flex-direction-column is-full px-5 mt-4">
+        <p>
+          This action will remove {contact.address} {contact.name ? `(${contact.name}) ` : ''}
+          from your saved addresses.
+        </p>
+      </div>
+      <div className="is-flex is-align-items-center mt-5 px-5">
+        <button className="button flex-1 p-4 mr-2" onClick={() => modalContext.closeModal()}>
+          Cancel
+        </button>
+        <button className="button is-link flex-1 p-4" onClick={() => {
+          modalContext.closeModal();
+          onConfirm();
+        }}>
+          {confirmText}
+        </button>
+      </div>
+    </div>
+  );
+}
+
+function EmptyContacts({ openAddModal }) {
+  return (
+    <section className="section is-flex is-align-items-center has-background-white-ter mt-4 rounded-lg" style={{ height: "calc(100vh - 340px)" }}>
+      <div className="container is-flex is-flex-direction-column is-justify-content-center is-align-items-center has-text-black">
+        <h2 className="is-size-5">This safe doesn't have any saved addresses.</h2>
+        <button className="button is-link mt-4" onClick={openAddModal}>
+          Add Contact <Plus className="ml-2" />
+        </button>
+      </div>
+    </section>
+  );
+}
+
+function SafeContacts({ address }) {
+  const modalContext = useModalContext();
   const clipboard = useClipboard();
-  const verifiedSafeOwners = safeOwners.filter((so) => so.verified);
+  const { contacts, setContact, removeContact } = useContacts(address);
+
+  const minWidth = 120;
+  const addressWidth = 160;
+
+  const openAddModal = () => modalContext.openModal(
+    <EditContactModal 
+      headerTitle="Add Contact"
+      confirmText="Add"
+      onConfirm={(newContact) => setContact(contacts.length, newContact)}
+      contact={{ name: '', address: '' }} 
+    />
+  );
+
+  const openEditModal = (index, contact) => modalContext.openModal(
+    <EditContactModal 
+      headerTitle="Edit Contact"
+      confirmText="Update"
+      onConfirm={(newContact) => setContact(index, newContact)}
+      contact={contact} 
+    />
+  );
+
+  const openRemoveModal = (index, contact) => modalContext.openModal(
+    <RemoveContactModal 
+      headerTitle="Confirm"
+      confirmText="Confirm"
+      onConfirm={() => removeContact(index)}
+      contact={contact} 
+    />
+  );
+
   return (
     <>
-      <div className="column p-0 mt-5 is-flex is-full">
-        <h4 className="is-size-5">Saved addresses</h4>
+      <div className="column p-0 mt-5 is-flex is-align-items-center is-justify-content-space-between is-full">
+        <h4 className="is-size-5">Saved Addresses</h4>
+        {!isEmpty(contacts) &&
```

Comment:
I'm a little confused on the conditions here, would this work?
```
{isEmpty(contacts)? (<EmptyContacts/>) 
: (
      <AddContactButton/>
      <ContactList/>
)}
```

Reply:
The Add Contact in the top right only shows up after you've already added at least one contact. If you dont have any contacts it only shows up in the body underneath the "This safe doesnt have any saved addresses". 

The reason there is two checks for isEmpty(contacts) is purely the structure of the dom where the Add contact button is nested inside the always visible "Saved Addresses" masthead.